### PR TITLE
Let Mbed TLS enable `ENTROPY_GENERATOR` when there is an entropy driver available on the platform

### PR DESCRIPTION
**Pull request by valeriosetti**
## TL;DR

This is an alternative implementation of #82497 since this PR had issues and it was reverted in #82807.

Here Mbed TLS enables `ENTROPY_GENERATOR` on platforms that have an entropy driver available. In order to do this a new Kconfig is added, `CONFIG_CSPRNG_AVAILABLE`, which is set to true when the devicetree chosen property "zephyr,entropy" is set.
When the entropy driver is not available, then Mbed TLS enables `CONFIG_TEST_RANDOM_GENERATOR` instead.

### Enabling the entropy driver

By default when a platform has an entropy HW available in the SOC it does _not_ get automatically enabled, but `CONFIG_ENTROPY_GENERATOR` has to be defined. The reason for this is that:
- in the devicetree we have both the node specific to the entropy device and "zephyr,entropy" chosen property set;
- in Kconfig we have the driver specific Kconfig (ex: `CONFIG_ENTROPY_NRF5_RNG`) enabled.

Since `CONFIG_ENTROPY_GENERATOR` is not enabled at "high level", the sub-Kconfig platform specific driver is not taken into account in the build so the driver is not enabled. This means that both `CONFIG_ENTROPY_HAS_DRIVER` and `CONFIG_CSPRNG_ENABLED` are not enabled by default.

### Mbed TLS

Mbed TLS need a cryptographically secure random number generator (CSPRNG) to work properly from a crypto security point of view (key generation, signatures, etc). A test/fake random generator would work the same from a build point of view, but that would be extremely bad on the practical side because all crypto algorithms would become extremely weak and easily to attack.

## Proposed solution

Since `CONFIG_ENTROPY_HAS_DRIVER` and `CONFIG_CSPRNG_ENABLED` are _not_ enabled by default even on platforms that actually has an entropy IP there was no easy way for Mbed TLS to tell if `CONFIG_ENTROPY_GENERATOR` had to be enabled or not. What we can do is to add another help symbol, `CONFIG_CSPRNG_AVAILABLE`, which can be used to say something like: "if you enable `CONFIG_ENTROPY_GENERATOR` then you can achieve `CONFIG_CSPRNG_ENABLED` on this platfom".
Then when `CONFIG_CSPRNG_AVAILABLE` is not enabled, Mbed TLS "knows" that it has no other choice than enabling `TEST_RANDOM_GENERATOR` in order to be buildable. This is a bad implementation form a cryptographic point of view, there would be no better choice for such a platform a part from completely excluding all samples/tests that include crypto, of course.

## History (How do we came here?)

- In the TinyCrypt deprecation PR #79931 there were many boards/samples/tests changes which were simply aimed at enabling `CONFIG_ENTROPY_GENERATOR` to let Mbed TLS access some good entropy/random value source. Of course this was not really scalable.
- One possible option was that the subsystem willing to make use of Mbed TLS had to enable `CONFIG_ENTROPY_GENERATOR` on its own, but that was not really elegant either because that subsystem (ex: bluetooth) might be platform agnostic and does not really care about how that stuff is implemented.
- #82497 tried to overcome this problem by letting `CONFIG_ENTROPY_GENERATOR` auto-enabling whenever "zephyr,entropy" was set in the devicetree. However this caused a lot of side effects because entropy drivers started to be included in a lot of samples/tests which required ad-hoc fixes. That's why the PR was eventually reverted in #82807.

### CODE REVIEW

File: modules/mbedtls/Kconfig.tls-generic [hunk 1]
```diff
@@ -479,14 +479,22 @@ config MBEDTLS_SSL_EXTENDED_MASTER_SECRET
 	  which ensures that master secrets are different for every
 	  connection and every session.
 
+DT_CHOSEN_Z_ENTROPY := zephyr,entropy
```
Comment (jhedberg) [inline modules/mbedtls/Kconfig.tls-generic:482]: This doesn't seem to be used anywhere?

Comment (valeriosetti) [reply modules/mbedtls/Kconfig.tls-generic:482]: Ops, sorry, that's a leftover from the devlopment

**Review by jhedberg — DISMISSED** [review]

File: soc/nxp/imxrt/Kconfig.defconfig [hunk 2]
```diff
@@ -82,8 +82,7 @@ choice SEGGER_SYSVIEW_SECTION
 endchoice
 
 config MBEDTLS
-	default y if CSPRNG_ENABLED
-	depends on ENTROPY_GENERATOR
+	default y if CSPRNG_AVAILABLE
```
Comment (dleach02) [inline soc/nxp/imxrt/Kconfig.defconfig:85]: I wonder if it wouldn't be better to not drive configuration of MBEDTLS based on the setting of CSPRNG.

When I added this, part of the reason you see these in the NXP parts is that we also had situations where we were trying to select with crypto library to use. I think though that we shouldn't automatically enable MBEDTLS but instead leave that to the subsystems and applications.

Comment (valeriosetti) [reply soc/nxp/imxrt/Kconfig.defconfig:85]: > I think though that we shouldn't automatically enable MBEDTLS but instead leave that to the subsystems and applications.

I agree. Now that the CI is OK, I can add more commits where I remove stuff like the one you proposed here. For example there is also [this change](https://github.com/zephyrproject-rtos/zephyr/blob/fcbf70f44d3c6d072288f8505bb1335faabdfb94/boards/nordic/nrf5340_audio_dk/Kconfig.defconfig#L63) that I introduced in #79931 which should no more needed now.

**Review by jhedberg — DISMISSED** [review]

File: subsys/random/Kconfig [hunk 3]
```diff
@@ -75,6 +75,19 @@ config XOSHIRO_RANDOM_GENERATOR
...
 endchoice # RNG_GENERATOR_CHOICE
 
+
+DT_CHOSEN_Z_ENTROPY := zephyr,entropy
+config CSPRNG_AVAILABLE
+	bool
+	default y if $(dt_chosen_enabled,$(DT_CHOSEN_Z_ENTROPY))
+	help
+	  Helper that can be used to check if the platform is capable of generating
+	  CS random values. For this to be enabled, there must be the "zephyr,entropy"
```
Comment (erwango) [inline subsys/random/Kconfig:85]: Complete nitpick:
CS acronym is introduced only below in the file. So I'd make it explicit in this help section.

**Review by erwango — COMMENTED** [review]
Shouldn't you re-introduce https://github.com/zephyrproject-rtos/zephyr/commit/a43cae0b3e40681def6f4a3385fe27e5c0e7ea6c ("doc: updates concerning ENTROPY_GENRATOR auto-enabling") along with this change ?

File: subsys/mgmt/osdp/Kconfig [hunk 4]
```diff
@@ -71,7 +71,7 @@ config OSDP_SKIP_MARK_BYTE
 
 config OSDP_SC_ENABLED
 	bool "OSDP Secure Channel"
-	depends on CSPRNG_ENABLED
+	depends on CSPRNG_AVAILABLE
```
Comment (dleach02) [inline subsys/mgmt/osdp/Kconfig:74]: The weirdness here is that we have this set of Kconfigs mixing around the RNG and CSPRNG enablement and needs. 

I was also playing around with a "CSPRNG_AVAILABLE" type of setting but what I think we really need is a "CSPRNG_ENABLE" and "RNG_ENABLE". Because right now we are using the ENABLE_GENERATOR as basically the same configuration which doesn't make sense. If a subsystem/application knows it needs CSRNG or RNG then just set the ENABLE for it and let the Kconfig dependency change get automatically set.

Comment (valeriosetti) [reply subsys/mgmt/osdp/Kconfig:74]: > If a subsystem/application knows it needs CSRNG or RNG then just set the ENABLE for it and let the Kconfig dependency change get automatically set.

I don't know if this would solve the problem I addressed with `CSPRNG_AVAILABLE`. As I wrote in the description this work was started by the fact that subsystems maintainers (bluetooth mainly @jhedberg ) were not happy about having to enable `ENTROPY_GENERATOR` (which would be `CSPRNG_ENABLE` if we follow your suggestion) on every board. As far as I understood they just want to say something like "I need a random generator and crypto support and I don't care about how that's implemented in the platform as long as both work properly". This summarizes as "let Mbed TLS enable what it needs in order to get the features that I asked".
The problem is that various subsystems are then tested on a large number of platforms, some of which does not have any capability to generate CSPRNG, but they might still be interested in running the sample/test. For example I just fixed an issue on `zephyr/samples/psa/its/sample.psa.its.secure_storage.entropy_not_secure` which intentionally tests secure storage on platforms that do not have an entropy source. So Mbed TLS "needs to know" what it must enable in order to build properly (is it `ENTROPY_GENERATOR` or `TEST_RANDOM_GENERATOR`?).

Setting `CSPRNG_ENABLE` directly (instead of having it indirectly selected) is another approach and I suspect that this might require us to rework all the samples/tests that then would fail to build. As of now instead this PR does not change any of those (albeit I admit that the work is not 100% done yet, so this might change).

Comment (dleach02) [reply subsys/mgmt/osdp/Kconfig:74]: I'm not implying doing this on every board. What I'm implying is that every subsystem/application that needs RNG would set directly that it requires CSPRNG or RNG. Let the underlying Kconfig infrastructure handle setting what is needed to accomplish this. 

This is what I don't like about the current system. To get CSPRNG you have to set ENTROPY_GENERATOR... which only makes sense to us doing the entropy or low level RNG stuff.

Comment (valeriosetti) [reply subsys/mgmt/osdp/Kconfig:74]: > This is what I don't like about the current system. To get CSPRNG you have to set ENTROPY_GENERATOR

Ok, I see and I also agree that this is not a user friendly. We can definitely improve this.

> What I'm implying is that every subsystem/application that needs RNG would set directly that it requires CSPRNG or RNG. Let the underlying Kconfig infrastructure handle setting what is needed to accomplish this.

I agree also on this, with only a minor note. IMO if we really want to make it generic, we should just add the "enable RNG" Kconfig symbol without specifying explicitly if it's CS or not, then under the hood we do the best we can (i.e. enable CS is there is any way to achieve it, otherwise non-CS). The reason for this is: who would prefer a non-CS source when there is a CS one that claims to work better for the same platform?

Comment (ceolin) [reply subsys/mgmt/osdp/Kconfig:74]: > I agree also on this, with only a minor note. IMO if we really want to make it generic, we should just add the "enable RNG" Kconfig symbol without specifying explicitly if it's CS or not, then under the hood we do the best we can (i.e. enable CS is there is any way to achieve it, otherwise non-CS). The reason for this is: who would prefer a non-CS source when there is a CS one that claims to work better for the same platform?

I prefer make these options explicit, there are different use cases for rng and csrng. Also if forces user to be thoughtful about the choice.

I would go beyond that, cryptography requires csrng and should enable it, if the target is not capable of that is up the to user to enable a non-cs rng. This would mean in our case that we enable csrng with MBEDTLS (I agree with @jhedberg  subsystem shouldn't need to figure out crypto dependencies by their own, they just want to have crypto ! ), if there is no CSPRNG available the application is responsible to enable a non-csrng by their own.

I don't think we should make exceptions about requirement, if we need CSRNG we need CSRNG. It is up to the user to workaround it because there are security consequences.

Comment (ceolin) [reply subsys/mgmt/osdp/Kconfig:74]: Ok, looking it more now I think the root cause of the problem is in the random subsys symbols, there is a lot of mixed things between entropy and randomness. Users / applications usually should care about randomness (csrng x rng) and not need to deal with entropy source. Let's discuss / brainstorm it more carefully in the security working group (12/16/2024).

**Review by jhedberg — DISMISSED** [review]

File: subsys/random/Kconfig [hunk 5]
```diff
@@ -75,6 +75,19 @@ config XOSHIRO_RANDOM_GENERATOR
 
 endchoice # RNG_GENERATOR_CHOICE
 
+
+DT_CHOSEN_Z_ENTROPY := zephyr,entropy
+config CSPRNG_AVAILABLE
+	bool
+	default y if $(dt_chosen_enabled,$(DT_CHOSEN_Z_ENTROPY))
```
Comment (frkv) [inline subsys/random/Kconfig:82]: Cleanliness-question:

Should we touch the device-tree here or should there be a generation of a prompt-less symbol called `DT_HAS_ENTROPY_ENABLED` somewhere in the system, instead?

I did a search for dt_chosen_enabled in Kconfigs (`zephyr/**/Kconfig*`) and I only find 19 entries... I find 3700 hits for `DT_HAS_XXXX_ENABLED` in Kconfig files

Comment (valeriosetti) [reply subsys/random/Kconfig:82]: Checking the `build/zephyr/include/generated/zephyr/autoconf.h` header file that is generated during the build and I did not find any `DT_HAS` for the chosen symbols.

For example:
`west build -b nrf52840dk/nrf52840 samples/hello_world/`
has
- `zephyr,entropy = &rng;` where `rng: random@4000d000` has status "okay" in `zephyr/build/zephyr/zephyr.dts` and
- `CONFIG_DT_HAS_NORDIC_NRF_RNG_ENABLED=y` in `build/zephyr/include/generated/zephyr/autoconf.h`

but no reference to chosen devicetree chosen property.

If we want to use DT_HAS I can modify `CSPRNG_AVAILABLE` as follows:
```
config CSPRNG_AVAILABLE
     bool
     default y
     depends on CONFIG_DT_HAS_NORDIC_NRF_RNG_ENABLED ||
                         CONFIG_DT_HAS_...._ENABLED ||
                         etc etc
```
I think the result would be the same :)

Comment (jhedberg) [reply subsys/random/Kconfig:82]: @frkv The `DT_HAS_XXXX_ENABLED` symbols are automatically derived from the compat strings of DT nodes which are enabled. This is not related to the chosen node, and there's also no single compat string for entropy drivers (they all have their own). It's a common convention to set driver Kconfig dependencies on the existence of an enabled node with the right compat string, which is why you're seeing all of those occurrences, but I don't think it helps us in this situation since this is about a property on the chosen node.

Comment (frkv) [reply subsys/random/Kconfig:82]: Thanks for the explanation @jhedberg. This is only nitpicking from me in the current PR at this point and I'm sure it can be considered for follow-up work if there needs to be a generic signal, or if we need to do larger structural changes to simplify things even further....

**Review by frkv — APPROVED** [review]
LGTM

Simplifications related to random subsystem signals and dependency of entropy symbol is likely too large a scope for this PR. I hope we can tackle this in follow-up work instead...

**Comment by valeriosetti** [comment]
As reported by @dleach02 the fix introduced here is still failing in:
`west build -p -b frdm_k64f/mk64f12 samples/bluetooth/peripheral_ht -T sample.bluetooth.peripheral_ht.frdm_kw41z_shield`

**Comment by jhedberg** [comment]
> As reported by @dleach02 the fix introduced here is still failing in:
> `west build -p -b frdm_k64f/mk64f12 samples/bluetooth/peripheral_ht -T sample.bluetooth.peripheral_ht.frdm_kw41z_shield`

Is it really? I just checked out this PR branch and ran the above command, and it seems to build just fine.

**Comment by valeriosetti** [comment]
I had to rebase to solve a conflict with `main`, but I did not change anything during the rebase so all the commits belonging to this PR are the same.

**Review by jhedberg — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Comment by jhedberg** [comment]
@valeriosetti there's another conflict in `samples/bluetooth/st_ble_sensor/prj.conf` that needs fixing.

After that, @ceolin @d3zd3z please have a look.

**Review by ceolin — DISMISSED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by ceolin — APPROVED** [review]
